evm_arithmetization/src/verifier.rs

```diff
@@ -15,9 +20,75 @@ use crate::cpu::kernel::aggregator::KERNEL;
 use crate::cpu::kernel::constants::global_metadata::GlobalMetadata;
 use crate::memory::segments::Segment;
 use crate::memory::VALUE_LIMBS;
-use crate::proof::{AllProof, AllProofChallenges, PublicValues};
+use crate::proof::{AllProof, AllProofChallenges, MemCap, PublicValues};
 use crate::util::h2u;
 
+pub(crate) fn initial_memory_merkle_cap<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    rate_bits: usize,
+    cap_height: usize,
+) -> MerkleCap<F, C::Hasher> {
+    // At the start of a transaction proof, `MemBefore` only contains the RLP
+    // constant and the `ShiftTable`.
+    let mut trace = Vec::with_capacity((KERNEL.code.len() + 256).next_power_of_two());
+
+    // Push kernel code.
+    for (i, &byte) in KERNEL.code.iter().enumerate() {
+        let mut row = vec![F::ZERO; crate::memory_continuation::columns::NUM_COLUMNS];
+        row[crate::memory_continuation::columns::FILTER] = F::ONE;
+        row[crate::memory_continuation::columns::ADDR_CONTEXT] = F::ZERO;
+        row[crate::memory_continuation::columns::ADDR_SEGMENT] =
+            F::from_canonical_usize(Segment::Code.unscale());
+        row[crate::memory_continuation::columns::ADDR_VIRTUAL] = F::from_canonical_usize(i);
+        row[crate::memory_continuation::columns::value_limb(0)] = F::from_canonical_u8(byte);
+        trace.push(row);
+    }
+    // Push shift table.
+    for i in 0..256 {
+        let mut row = vec![F::ZERO; crate::memory_continuation::columns::NUM_COLUMNS];
+        let val = U256::from(1) << i;
+        row[crate::memory_continuation::columns::FILTER] = F::ONE;
+        row[crate::memory_continuation::columns::ADDR_CONTEXT] = F::ZERO;
+        row[crate::memory_continuation::columns::ADDR_SEGMENT] =
+            F::from_canonical_usize(Segment::ShiftTable.unscale());
+        row[crate::memory_continuation::columns::ADDR_VIRTUAL] = F::from_canonical_usize(i);
+        for j in 0..crate::memory::VALUE_LIMBS {
+            row[j + 4] = F::from_canonical_u32((val >> (j * 32)).low_u32());
+        }
+        trace.push(row);
+    }
+
+    // Padding.
+    let num_rows = trace.len();
+    let num_rows_padded = num_rows.next_power_of_two();
+    for _ in num_rows..num_rows_padded {
+        trace.push(vec![
+            F::ZERO;
+            crate::memory_continuation::columns::NUM_COLUMNS
+        ]);
+    }
```

Comment:
may be better to extend from an iterator, but no strong opinion

Reply:
I used `resize` instead.